panel/widgets/select.py

```diff
@@ -654,6 +673,73 @@ def _update_options_programmatically(self):
                 self.value = original_values
                 raise
 
+    @classmethod
+    def create_options_from_dataframe(cls, data: pd.DataFrame, columns: List|None=None)->Dict|List:
+        """Returns the options to use with a NestedSelect
+
+        Args:
+            data (pd.DataFrame): A DataFrame to create options from.
+            columns (List | None, optional): The columns to create options from. If None then all
+                columns are used. Defaults to None.
+
+        Returns:
+            Dict|List: A dictionary (or list) of options. The options are sorted ascending.
+        """
+        if columns is None:
+            columns = list(data.columns)
+
+        data = data[columns].drop_duplicates().sort_values(columns).reset_index(drop=True)
+        options = _build_nested_dict(data)
+        return options
+
+    @classmethod
+    def create_from_dataframe(cls, data: pd.DataFrame, columns: List|None=None, **params)->'NestedSelect':
+        """Returns a NestedSelect with options from the data and columns specified
+
+        Args:
+            data (pd.DataFrame): A DataFrame to create options from.
+            columns (List | None, optional): The columns to create options from. If None then all
+                columns are used. Defaults to None. If level is not provided the columns are used as
+                levels.
+            params: Any parameters (besides options) that the NestedSelect accepts
+
+        Returns:
+            NestedSelect: A NestedSelect with options from the data and columns specified
+        """
+        if not columns:
+            columns = list(data.columns)
+
+        options = cls.create_options_from_dataframe(data, columns)
+        params["levels"]=params.get("levels", columns)
+        return NestedSelect(options=options, **params)
+
+    @classmethod
+    def filter_dataframe(cls, data: pd.DataFrame, value: Dict, columns: List|None=None, **params)->pd.DataFrame:
+        """Returns the subset of the data corresponding to the filter given by the value
+
+        Args:
+            data (pd.DataFrame): The original data to filter
+            value (Dict): A dictionary of key, value filter options
+            columns (List | None, optional): An optional list of column names to use if the keys of
+                the value do not correspond to the column names. Defaults to None.
+
+        Returns:
+            pd.DataFrame: The filtered subset
+        """
+        if data.empty or not value:
+            return data
+
+        if columns:
+            value = {new_key: value for new_key, value in zip(columns, value.values())}
+
+        import pandas as pd
```

Comment:
Should I just import pandas at top of module?

Reply:
Importing pandas takes time, and we don't want to do it unless you have to use it. So, what you have done here is correct.

Reply:
I think it should at least at the start of the function.

Reply:
Here is fine, if moved to the start of the function you'd import it even if it wasn't needed.